### PR TITLE
added function url resource to module

### PR DESCRIPTION
**Pull request by Halima-RV**

### CODE REVIEW

File: examples/example-with-functionurl/main.tf [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  name = "tf-example-go-basic"
+  handler       = "example-lambda-func"
+  runtime       = "go1.x"
+  service       = "example"
+  project       = "example"
+  environment   = "qa"
+  team_name     = "example"
+  owner         = "example"
+  enable_functionurl = true
+}
```
Comment (rv-jgentile) [inline examples/example-with-functionurl/main.tf]: Mind running a quick `terraform fmt` on this dir real quick just to line these operators up?

File: variables.tf [hunk 2]
```diff
@@ -27,6 +27,17 @@ variable architecture {
...
+  type = string
+  description = "Alias name or $LATEST"
+  default = "$LATEST"
+}
+
+variable "enable_functionurl" {
+  type = bool
+  description = "trigger to create lambda function url"
+  default = false
+}
```
Comment (rv-jgentile) [inline variables.tf]: `terraform fmt` on this dir as well pls

**Comment by rv-jgentile** [comment]
Let's include `aws_lambda_function_url` in the outputs here. Might take a bit of trickery since it may or may not be created, but it can probably be addressed with a ternary that does (pseudo-code incoming):
```
value = enable_functionurl ? aws_lambda_function_url : ""
```
shouldn't take too much tinkering to get it working
  
  _Edit: Just realized I should probably include reasoning, lol._ 
  
Outputting the url will make it a lot easier for users. If they want a function url then they're gonna want to use it _somewhere_. Adding it to outputs makes it obvious what url was created. Without that output they'd probably have to go digging around the console to figure out what endpoint they have to plug into their app/browser/etc. Might as well just throw it up on the screen.

File: variables.tf (hunk 2 above)
Comment (Halima-RV) [reply variables.tf]: Done, Thank you

**Review by rv-jgentile — DISMISSED** [review]

**Comment by lbrimeyer** [comment]
> Let's include `aws_lambda_function_url` in the outputs here. Might take a bit of trickery since it may or may not be created, but it can probably be addressed with a ternary that does (pseudo-code incoming):
> 
> ```
> value = enable_functionurl ? aws_lambda_function_url : ""
> ```
> 
> shouldn't take too much tinkering to get it working
> 
> _Edit: Just realized I should probably include reasoning, lol._
> 
> Outputting the url will make it a lot easier for users. If they want a function url then they're gonna want to use it _somewhere_. Adding it to outputs makes it obvious what url was created. Without that output they'd probably have to go digging around the console to figure out what endpoint they have to plug into their app/browser/etc. Might as well just throw it up on the screen.

```
value = enable_functionurl ? aws_lambda_function_url : null
```
This would be my preference if it works to avoid outputting anything if it's disabled. 
Worth noting in a ton of languages (ideally one day including HCL/GoLang) null != undefined != empty/zero-value and the difference in those meanings is usually important.

File: main.tf [hunk 3]
```diff
@@ -152,5 +152,11 @@ resource "aws_iam_role_policy_attachment" "kms_policy_attachment" {
   count      = var.kms_key_arn != "" ? 1 : 0
   role       = module.lambda.iam_role_name
   policy_arn = aws_iam_policy.kms_policy[count.index].arn
-} 
+}
 
+resource "aws_lambda_function_url" "lambda_url" {
+  count              = var.enable_functionurl ? 1 : 0
```
Comment (lbrimeyer) [inline main.tf]:
```suggestion
  count              = var.enable && lookup(var.architecture, "function_url", false) ? 1 : 0
```
This follow the architecture pattern we've established for the other event sources and lambda triggers.

File: variables.tf [hunk 4]
```diff
@@ -1,32 +1,43 @@
...
   })
 
   default = {
-    cloudwatch_trigger             = false
-    s3_trigger                     = false
-    ddb_trigger                    = false
+    cloudwatch_trigger = false
+    s3_trigger         = false
+    ddb_trigger        = false
   }
```
Comment (lbrimeyer) [inline variables.tf]: I'd add function_url to this struct.

**Review by lbrimeyer — CHANGES_REQUESTED** [review]
Minor structure change for architecture to include function_url instead of an entirely new variable and how the output is handled.

File: main.tf [hunk 5]
```diff
@@ -152,5 +152,10 @@ resource "aws_iam_role_policy_attachment" "kms_policy_attachment" {
   count      = var.kms_key_arn != "" ? 1 : 0
   role       = module.lambda.iam_role_name
   policy_arn = aws_iam_policy.kms_policy[count.index].arn
-} 
+}
 
+resource "aws_lambda_function_url" "lambda_url" {
+  count              = var.enable && lookup(var.architecture, "function_url", false) ? 1 : 0
+  function_name      = module.lambda.arn
+  authorization_type = "AWS_IAM"
```
Comment (lbrimeyer) [inline main.tf]: Let's swap this to a variable that validates whether either `AWS_IAM` or `NONE`,  default to `AWS_IAM` as the default (more secure) option out of the gate. Great points during Zoom PR 🏅 .

**Review by lbrimeyer — CHANGES_REQUESTED** [review]
Adding the variable for auth type but otherwise this looks slick!

File: variables.tf [hunk 6]
```diff
@@ -174,23 +176,29 @@ variable "timeout" {
...
 variable "layers" {
   description = "List of Lambda Layer Version ARNs (maximum of 5) to attach to your Lambda Function. See [Lambda Layers](https://docs.aws.amazon.com/lambda/latest/dg/configuration-layers.html)"
   type        = list(string)
   default     = []
 }
+
+variable "authorization_type" {
+  description = "The type of authentication that the function URL uses. Defaults to AWS_IAM which restricts access to authenticated users. Use NONE to allow public unauthenticated users invoke your lambda"
+  type = string
+  default = "AWS_IAM"
```
Comment (lbrimeyer) [inline variables.tf:203]:
```suggestion
  default = "AWS_IAM"
  
  validation {
    condition     = contains(["AWS_IAM", "NONE"], var.authorization_type)
    error_message = "Authorization type must be either `AWS_IAM` || `NONE`"
  }
```

**Review by lbrimeyer — CHANGES_REQUESTED** [review]
Super super close 🏅

**Review by lbrimeyer — APPROVED** [review]

**Review by rv-jgentile — APPROVED** [review]

**Pull request merged**